Death markers setting

OhmV-IR opened the pull request:
The death markers setting places beacons where players die so they can retrieve items or equipment that was left behind. 
I have tested the PR in-game using 2 players, swim out a bit from the lifepod and use the kill command or suffocate, beacon should only appear for the dying player. Then when that player goes within 20m of the beacon or disconnects from the server the beacon should disappear. The beacon can also be hidden from the HUD through the PDA. Ready for CW reviews.

## Closes
- Closes https://github.com/SubnauticaNitrox/Nitrox/issues/2155

OhmV-IR committed:
Create beacons on all clients

OhmV-IR committed:
Fix syntax error

OhmV-IR committed:
Shift to using the existing PlayerDeathEvent packet to trigger the beacon spawns, beacon spawn code working

OhmV-IR committed:
Make death beacons player exclusive(only spawns for the player that triggers them)

OhmV-IR committed:
Disconnecting also despawns all active death beacons

OhmV-IR committed:
Fix death beacons getting synced to other players and not disappearing on disconnect

OhmV-IR committed:
Remove unnecessary logging

OhmV-IR committed:
Remove unnecessary method, various semantic updates to prep for PR

Coding-Hen commented:
I haven't tested this yet but a general question of does this still try and spawn a beacon even if the person is on land, if this is desired then does it spawn and roll around potentially not being in the correct place

Coding-Hen commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        this.playerManager = playerManager;
+        this.serverConfig = serverConfig;
+        AddParameter(new TypeBoolean("state", true, "the on/off state to set death markers to"));
```
```suggestion
        AddParameter(new TypeBoolean("state", true, "the on/off state of if a death marker is spawned on death"));
```

Coding-Hen commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,44 @@ public override void Process(PlayerDeathEvent playerDeath)
+            Pickupable item = beacon.GetComponent<Pickupable>();
+            if (item != null)
+            {
+                using (PacketSuppressor<EntitySpawnedByClient>.Suppress()) // Prevent the beacon from syncing to other players or the server so it will disappear on disconnect
```
I would like to think that we could make this sync between sessions but only for that person but maybe that is for another ticket after we have added basic functionality

OhmV-IR replied:
Yeah for it to be saved the server would have to know about it and the server would replicate it to the other players. I'd rather save that can of worms for another time

Coding-Hen commented on NitroxClient/Communication/MultiplayerSession/ConnectionState/CommunicatingState.cs:
```diff
@@ -1,5 +1,6 @@
-﻿using System.Threading.Tasks;
+using System.Threading.Tasks;
 using NitroxClient.Communication.Abstract;
+using NitroxClient.Communication.Packets.Processors;
```
Did we need the changes in this file?

OhmV-IR replied:
Whoops was looking through files and accidentally added an import, fixed

OhmV-IR committed:
Update NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs

Co-authored-by: Coding-Hen <8798074+Coding-Hen@users.noreply.github.com>

OhmV-IR committed:
Fixed a random import that was accidentally added

OhmV-IR commented:
Yeah so when it spawns on land it actually falls through the land to the water below. To combat this, I set the gravitation constant in the beacon's WorldForces component to 0 so gravity has no effect on it and added above-water drag because floating point error gave the object a tiny velocity in a random direction causing it to drag. it works on land now. Tested in-water again too

OhmV-IR committed:
Fixed deathbeacons on land

OhmV-IR committed:
Switch to using blank gameobjects instead of actual beacons

OhmV-IR commented:
Instead of using beacons have just switched to using blank gameobjects and attaching the PingInstance component. tested IG again

Jannify commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
+
+public class DeathBeacon : MonoBehaviour
+{
+    private static readonly float despawnDistance = 20f;
```
Should be const

OhmV-IR replied:
Implemented

Jannify commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
+
+    public static IEnumerator SpawnDeathBeacon(NitroxVector3 location, string playerName)
+    {
+        GameObject beacon = new();
```
Would give the GameObject a name

Jannify commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
     }
 }
+
+public class DeathBeacon : MonoBehaviour
```
Should be it's own file inside `NitroxClient.MonoBehaviours.Gui.MainMenu`

OhmV-IR replied:
Made this into its own file in NitroxClient.MonoBehaviours.Gui.InGame which is more fitting than the MainMenu folder, since these are only used in-game. Cleaned and sorted usings on PlayerDeathProcessor.cs and LocalPlayer.cs

Jannify commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
+{
+    private static readonly float despawnDistance = 20f;
+
+    public static IEnumerator SpawnDeathBeacon(NitroxVector3 location, string playerName)
```
Why is this a Coroutine?

Jannify commented on NitroxClient/GameLogic/LocalPlayer.cs:
```diff
@@ -37,10 +39,10 @@ public class LocalPlayer : ILocalNitroxPlayer
     public Perms Permissions { get; set; }
     public IntroCinematicMode IntroCinematicMode { get; set; }
     public bool KeepInventoryOnDeath { get; set; }
+    public bool MarkDeathPointsWithBeacon { get; set; }
```
I think we should put variables like this (world settings per player) in they own container class and update

OhmV-IR replied:
I really don't think the current system hurts anything, and considering how IntroCinematicMode qualifies as a "world setting per player" and is under active development in another PR(https://github.com/SubnauticaNitrox/Nitrox/pull/2257), I'd like to hold off on doing this and leave it for a later refactor or something. Won't open a GH issue just yet though

tornac1234 replied:
Actually there "could" be a bool for PvPEnabled but I decided to let this on server-side only. But we could definitely end up making a "ServerSettings" wrapper on the client which would be more relevant than the current model in which those variables seem to caracterize the player itself. But I agree that it's not an urgent fact and it can be done later on

Jannify commented on NitroxPatcher/Patches/Dynamic/Player_OnKill_Patch.cs:
```diff
@@ -1,17 +1,24 @@
 using System.Reflection.Emit;
 using HarmonyLib;
+using NitroxClient.Communication.Abstract;
+using NitroxClient.GameLogic;
+using NitroxClient.GameLogic.Spawning.WorldEntities;
+using NitroxModel.DataStructures;
+using NitroxModel.DataStructures.Unity;
 using NitroxModel.Helper;
-
+using NitroxModel.Packets;
+using UnityEngine;
```
What?

OhmV-IR replied:
Cleaned and sorted usings on the file, should fix the extra changes this was a first implementation of the setting, but later I pivoted to using the event built into LocalPlayer which is simpler

Jannify commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,40 @@
+using NitroxModel.Packets;
+using System.IO;
+using NitroxServer.ConsoleCommands.Abstract;
```
Wrong sorting

OhmV-IR replied:
Reran clean and sort usings on this file

Jannify commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace NitroxServer.ConsoleCommands;
+
+internal class SetDeathMarkersCommand : Command
```
```suggestion
public class SetDeathMarkersCommand : Command
```

OhmV-IR replied:
Every other command is an internal class, is there a reason why the death markers command should be different?

dartasen replied:
internal is fine

Jannify commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,40 @@
+    private readonly PlayerManager playerManager;
+    private readonly ServerConfig serverConfig;
+
+    public SetDeathMarkersCommand(PlayerManager playerManager, ServerConfig serverConfig) : base("setdeathmarkers", NitroxModel.DataStructures.GameLogic.Perms.ADMIN, "Sets \"Death Markers\" setting to on/off. If \"on\", a beacon will appear at the location where a player dies.")
```
You should add `using NitroxModel.DataStructures.GameLogic;`

OhmV-IR replied:
Implemented

OhmV-IR committed:
Added missing using

OhmV-IR committed:
Removed unneeded changes

OhmV-IR committed:
Moved DeathBeacon class to its own file

OhmV-IR committed:
changed field to const and named deathbeacon GO

OhmV-IR committed:
Use InvokeRepeating and sqrMagnitude to check distances for better performance

OhmV-IR commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
+
+    private void Update()
+    {
+        if(Vector3.Distance(Player.main.transform.position, transform.position) <= despawnDistance)
```
Implemented the InvokeRepeating and sqrMagnitude solution

OhmV-IR commented on NitroxClient/Communication/Packets/Processors/PlayerDeathProcessor.cs:
```diff
@@ -20,7 +29,33 @@ public override void Process(PlayerDeathEvent playerDeath)
+        signal.pingType = PingType.Signal;
+        signal.displayPingInManager = true;
+        signal.origin = beacon.transform;
+        signal._label = $"{playerName}'s death";
```
Am now using Language.main.Get to fetch from the language files, have added the english string to the json file with the previous text, not using for the gameobject's name though

OhmV-IR committed:
Add translation string for death beacon label

Jannify commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+        PingInstance signal = beacon.AddComponent<PingInstance>();
+        signal.pingType = PingType.Signal;
+        signal.displayPingInManager = true;
+        signal.origin = beacon.transform;
```
You set this field after `beacon.transform.position = location.ToUnity();`.
Are you sure the view distance is handled correctly (when looking at `PingInstance.IsVisible()`)

OhmV-IR replied:
moved setting transform position of beacon to right after it is spawned

Jannify commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+    }
+    private void CheckPlayerDistance()
+    {
+        if ((Player.main.transform.position - transform.position).sqrMagnitude <= despawnDistanceSquared)
```
I just noticed that `PingInstance` has a `visitDistance` field. Have you researched if we can use that instead of custom logic? I fear that it isn't as straight forward (bc of UWE) so pls feel free to discuss your findings with me on discord before implementing a final solution.

OhmV-IR replied:
PingManager seems to have an onvisit event but since PingManager is a singleton class, any changes there would most likely propagate to all other beacons and I haven't even been able to find how that event is used in vanilla because I can't see a reference to it anywhere

OhmV-IR committed:
implement various code review requsted changes

OhmV-IR commented on Nitrox.Assets.Subnautica/LanguageFiles/en.json:
```diff
@@ -58,6 +58,7 @@
     "Nitrox_OK": "Ok",
     "Nitrox_OutOfDateClient": "Your Nitrox installation is out of date. Server: {serverVersion}, Yours: {localVersion}.",
     "Nitrox_OutOfDateServer": "The server runs an older version of Nitrox. Ask the server admin to upgrade the server or downgrade your Nitrox installation. Server: {serverVersion}, Yours: {localVersion}.",
+    "Nitrox_PlayerDeathBeaconLabel":  "{PLAYER}'s death'",
```
implemented

OhmV-IR commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+    private const float despawnDistanceSquared = despawnDistance * despawnDistance;
+    private const float checkRate = 10f; // in seconds
+
+    public static IEnumerator SpawnDeathBeacon(NitroxVector3 location, string playerName)
```
Implemented

OhmV-IR commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+{
+    private const float despawnDistance = 20f;
+    private const float despawnDistanceSquared = despawnDistance * despawnDistance;
+    private const float checkRate = 10f; // in seconds
```
Implemented

OhmV-IR commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+
+    private void Start()
+    {
+        InvokeRepeating("CheckPlayerDistance", 0, checkRate);
```
Implemented

OhmV-IR committed:
Fix error in building

OhmV-IR committed:
Add beacon fade out

OhmV-IR commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        if ((Player.main.transform.position - transform.position).sqrMagnitude <= despawnDistanceSquared)
+        {
+            Destroy(gameObject);
```
There is some vanilla behavior here that we can use to our advantage. setting the minDist of the beacon to the despawn distance + 10 or so will make the beacon fade out from the UI before the player gets within the 20m where it will be destroyed

Jannify commented on NitroxClient/GameLogic/LocalPlayer.cs:
```diff
@@ -12,6 +12,7 @@
 using NitroxModel_Subnautica.DataStructures;
 using UnityEngine;
 using UnityEngine.Rendering;
+using UWE;
```
Not needed anymore

Jannify commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,37 @@
+    private void Start()
+    {
+        InvokeRepeating("CheckPlayerDistance", 0, checkRate);
+    }
+    private void CheckPlayerDistance()
```
This is still open

OhmV-IR committed:
Fixed add and sort usings, as well as added a whitespace

Jannify reviewed (approved):
LGTM CW

Coding-Hen reviewed (approved):
LGTM, it will be good to have a marker for ease of finding the body again

Coding-Hen commented:
Might be good to squash down some of the commits where sensible

NinjaPedroX commented on NitroxServer/Serialization/ServerConfig.cs:
```diff
@@ -30,6 +30,9 @@ public class ServerConfig : NitroxConfig<ServerConfig>
         public bool KeepInventoryOnDeath { get; set; } = false;
 
+        [PropertyDescription("Places a beacon where players die")]
+        public bool MarkDeathPointsWithBeacon { get; set; } = false;
```
Shouldn't the default value for this be set to true? I feel like it'd be more useful that way than forcing people to turn the setting on.

tornac1234 replied:
I agree it should be true by default

OhmV-IR committed:
set DeathMarkers setting to true by default

dartasen commented:
@Coding-Hen Has this been tested IG ?

OhmV-IR committed:
Added boilerplate for DeathMarkers setting

OhmV-IR commented:
Rebased to solve merge conflicts

OhmV-IR commented:
Updated this for new launcher and tested IG. Needs new CW reviews because some changes were made and its possible I goofed the rebase but I don't think I did

OhmV-IR committed:
Fix DeathMarkers for NitroxLauncher(tested in-game for this commit)

Jannify commented on NitroxClient/GameLogic/LocalPlayer.cs:
```diff
@@ -88,6 +89,10 @@ public void BroadcastDeath(Vector3 deathPosition)
         if (PlayerId.HasValue)
         {
             packetSender.Send(new PlayerDeathEvent(PlayerId.Value, deathPosition.ToDto()));
+            if (MarkDeathPointsWithBeacon)
+            {
+                DeathBeacon.SpawnDeathBeacon(deathPosition.ToDto(), PlayerName);
+            }
```
Should be moved to `PlayerDeathBroadcaster.PlayerDeath()` imo

OhmV-IR committed:
Added missing namespace declaration

OhmV-IR committed:
Moved spawning of death beacon to PlayerDeath in PlayerDeathBroadcaster

OhmV-IR commented:
Tested in-game on commit 99d9ce4 (Moved spawning of death meacon to PlayerDeath in PlayerDeathBroadcaster)

tornac1234 commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,42 @@
+        this.playerManager = playerManager;
+        this.serverConfig = serverConfig;
+        this.server = server;
+        AddParameter(new TypeBoolean("state", true, "the on/off state of if a death marker is spawned on death"));
```
```suggestion
        AddParameter(new TypeBoolean("state", true, "on/off to enable/disable death markers"));
```

tornac1234 commented on NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs:
```diff
@@ -0,0 +1,42 @@
+    private readonly SubnauticaServerConfig serverConfig;
+    private readonly Server server;
+
+    public SetDeathMarkersCommand(PlayerManager playerManager, SubnauticaServerConfig serverConfig, Server server) : base("setdeathmarkers", Perms.ADMIN, "Sets \"Death Markers\" setting to on/off. If \"on\", a beacon will appear at the location where a player dies.")
```
I think the command should be "deathmarkers" or "enabledeathmarkers" even though that's longer. Because "set" doesn't seem like the right verb to me. What's your guys mind on it ?

OhmV-IR replied:
Seeing as the command can be used to disable or enable the setting, enable definitely isn't the right verb. I'd be fine with either the current, since it is accurate to the function of setting the parameter, or just shortening it to deathmarkers

tornac1234 reviewed (changes requested):
Apart from those changes, a test session revealed the PR to be working (so we'll be able to merge it after that)

OhmV-IR committed:
Update NitroxModel/Packets/DeathMarkersChanged.cs

Co-authored-by: rootcan <24827220+tornac1234@users.noreply.github.com>

OhmV-IR committed:
Update NitroxServer/ConsoleCommands/SetDeathMarkersCommand.cs

Co-authored-by: rootcan <24827220+tornac1234@users.noreply.github.com>

OhmV-IR committed:
Add newline between namespace and class

OhmV-IR committed:
Merge branch 'DeathMarkers' of https://github.com/OhmV-IR/NitroxModdedUnofficial into DeathMarkers

OhmV-IR committed:
Fix missing import

OhmV-IR committed:
change setdeathmarkers to deathmarkers cmd name

Jannify reviewed (approved):
LGTM CW

tornac1234 reviewed (approved):
Command class name is not the right one but it's fine ig

Commit:
merge master

Commit:
Merge branch 'master' into DeathMarkers

Commit:
make it build

OhmV-IR commented:
made the merge but have not tested IG yet, gotta do that sometime

CherrieTheShifter commented:
Yes This!. Also This?, Anyways Sorry, Bye. GG
https://github.com/Sunrunner37/Nitrox

OhmV-IR committed:
Merge master

OhmV-IR committed:
Adapt to refactors (make it compile)

OhmV-IR committed:
fix logic error

OhmV-IR commented:
updated for latest master, tested ig and seems to still work. Checked both that the feature works IG and that it shows up correctly in the launcher.

Measurity committed:
DeathBeacon code cleanup

 - Grouped related instructions
 - Added "ToString" for ICommandContext returning OriginName and #OriginId
 - Used nameof where sensible to assist with future refactorings
 - Changed DeathBeacon despawning logic from distance check to physics trigger

Coding-Hen commented on NitroxClient/MonoBehaviours/Gui/InGame/DeathBeacon.cs:
```diff
@@ -0,0 +1,42 @@
+
+    public static void SpawnDeathBeacon(NitroxVector3 location, string playerName)
+    {
+        GameObject beacon = new($"{playerName}{nameof(DeathBeacon)}");
```
Will we have a problem with multiple having the same name or do we want to allow for easy identification if they die more than once?

OhmV-IR replied:
Yeah it might become a problem if players keep dying repeatedly trying to get their items from their previous death location. My proposed solution would be to add the time of the death IRL to the name of the beacon.
